lib/matplotlib/axes/_axes.py

```diff
@@ -2604,6 +2605,123 @@ def barh(self, y, width, height=0.8, left=None, *, align="center",
                            align=align, **kwargs)
         return patches
 
+    def bar_label(self, container, captions=[], *, fmt="%g", position="edge",
+                  padding=0, shifting=0, **kwargs):
+        """
+        Label a bar plot.
+
+        Adds labels to bars in the given `.BarContainer`.
+        You may need to adjust the axis limits to fit the labels.
+
+        Parameters
+        ----------
+        container : `.BarContainer`
+            Container with all the bars and optionally errorbars.
+
+        captions : array-like, optional
+            A list of label texts, that should be displayed. If not given, the
+            label texts will be the data values formatted with *fmt*.
+
+        fmt : str, default: '%g'
+            A format string for the label.
+
+        position : {'edge', 'center'}, default: 'edge'
+            Position of the label relative to the bar:
+
+            - 'edge': Placed at edge, the cumulative values will be shown.
+            - 'center': Placed at center, the individual values will be shown.
+
+        padding : float, default: 0
+            Offset in points for label to shift in longitudinal direction.
+
+        shifting : float, default: 0
+            Offset in points for label to shift in lateral direction.
+
+        Returns
+        -------
+        annotations
+            A list of `.Text` instances for the labels.
+        """
+        def extend_with_none(iterable, length):
+            return list(iterable) + [None] * (length - len(iterable))
+
+        def sign(x):
+            return 1 if x >= 0 else -1
+
+        cbook._check_in_list(['edge', 'center'], position=position)
+
+        bars = container.patches
+        errorbar = container.errorbar
+        orientation = container.orientation
+
+        N = len(bars)
+        if errorbar:
+            lines = errorbar.lines
+            barlinecols = lines[2]
+            barlinecol = barlinecols[0]
+            segments = barlinecol.get_segments()
+            errs = extend_with_none(segments, length=N)
+        else:
+            errs = extend_with_none([], length=N)
+        caps = extend_with_none(captions, length=N)
+
+        annotations = []
+        for bar, err, cap in zip(bars, errs, caps):
```

Comment:
Might be simpler to use `zip_longest` rather than manually extending all of the inputs?